### PR TITLE
hoc2015: smoothly fade out items & goals when touched

### PR DESCRIPTION
**Pull request by breville**

### CODE REVIEW

File: apps/src/studio/Item.js [hunk 1]
```diff
@@ -90,10 +94,193 @@ Item.prototype.createElement = function (parentElement) {
...
+/**
+ * This function should be called every frame, and moves the item around.
+ * It moves the item smoothly, but between fixed points on the grid.
+ * Each time the item reaches its destination fixed point, it reevaluates
+ * its next destination location based on the type of movement specified.
+ * It generally evalutes all possible destination locations, prioritizes
+ * the best possible moves, and chooses randomly between evenly-scored
+ * options.
+ */
+Item.prototype.update = function () {
```
Comment (breville) [inline apps/src/studio/Item.js:107]: This finally moves update() into item.js.

File: apps/src/studio/Item.js [hunk 2]
```diff
@@ -113,6 +300,25 @@ Item.prototype.removeElement = function () {
   Studio.trackedBehavior.removedItemCount++;
 };
 
+
+/**
+ * Returns true if the item is currently fading away.
+ */
+Item.prototype.isFading = function() {
+  return this.startFadeTime ? true : false;
```
Comment (cpirich) [inline apps/src/studio/Item.js]: also could use `!!this.startFadeTime`

File: apps/src/studio/Item.js [hunk 3]
```diff
@@ -113,6 +300,25 @@ Item.prototype.removeElement = function () {
...
+};
+
+
+/**
+ * Returns true if the item has finished fading away.  The caller will usually
+ * then call removeElement to destroy this item's assets.
+ */
+Item.prototype.isGone = function() {
```
Comment (cpirich) [inline apps/src/studio/Item.js]: might consider a more clear name like `hasCompletedFade` or something to that effect...

File: apps/src/studio/studio.js [hunk 4]
```diff
@@ -978,12 +978,14 @@ function edgeCollidableCollisionDistance (collidable, edgeName, yAxis) {
...
            spriteIndex, xCenter, yCenter, list, autoDisappear)
 {
-  // Traverse the list in reverse order because we may remove elements from the
-  // list while inside the loop:
-  for (var i = list.length - 1; i >= 0; i--) {
+  for (var i = 0; i < list.length; i++) {
     var collidable = list[i];
     var next = collidable.getNextPosition();
 
+    if (collidable.isFading()) {
```
Comment (cpirich) [inline apps/src/studio/studio.js]: This function is used for two types of `Collidable` lists: `Item` and `Projectile`. However, you are now calling two new methods (`isFading` and `beginRemoveElement`) that don't exist on `Collidable` or `Projectile`, but only exist on `Item`.

File: apps/src/studio/studio.js [hunk 5]
```diff
@@ -1016,8 +1018,7 @@ function handleActorCollisionsWithCollidableList (
         // Make the projectile/item disappear automatically if this parameter
         // is set:
         if (autoDisappear) {
-          collidable.removeElement();
-          list.splice(i, 1);
+          collidable.beginRemoveElement();
```
Comment (cpirich) [inline apps/src/studio/studio.js]: If you want to use this technique for projectiles as well, you will need to add a function similar to `updateItems` for projectiles so you can remove them when `isGone` returns `true`. Or you can go back to immediately calling `removeElement` for projectiles. You could check to see if the `beginRemoveElement` method exists. But if you do it conditionally, you'll need to flip the loop back to going backwards...

**Comment by cpirich** [comment]
Looks good, but there are some issues that will need to be addressed.

Also, does setting `skin.fadeOutGoal` mean that you'll never see a success state for the goal? If so, another approach would be to set `skin.goalSuccess` to null if you desire a fadeout instead of a different image. While that's not quite as obvious, the priority of `fadeOutGoal` over `goalSuccess` is not obvious either.

File: apps/src/studio/studio.js [hunk 6]
```diff
@@ -1016,8 +1020,11 @@ function handleActorCollisionsWithCollidableList (
         // Make the projectile/item disappear automatically if this parameter
         // is set:
         if (autoDisappear) {
-          collidable.removeElement();
-          list.splice(i, 1);
+          if (collidable.beginRemoveElement) {
+            collidable.beginRemoveElement();
+          } else {
+            collidable.removeElement();
```
Comment (cpirich) [inline apps/src/studio/studio.js:1026]: add back the splice call

**Pull request merged**